python/ray/monitor.py

```diff
@@ -142,16 +143,46 @@ def cleanup_task_table(self):
         num_tasks_updated = 0
         for task_id, task in tasks.items():
             # See if the corresponding local scheduler is alive.
-            if task["LocalSchedulerID"] in self.dead_local_schedulers:
-                # If the task is scheduled on a dead local scheduler, mark the
-                # task as lost.
-                key = binary_to_object_id(hex_to_binary(task_id))
-                ok = self.state._execute_command(
-                    key, "RAY.TASK_TABLE_UPDATE", hex_to_binary(task_id),
-                    ray.experimental.state.TASK_STATUS_LOST, NIL_ID)
-                if ok != b"OK":
-                    log.warn("Failed to update lost task for dead scheduler.")
-                num_tasks_updated += 1
+            if task["LocalSchedulerID"] not in self.dead_local_schedulers:
+                continue
+
+            # Remove dummy objects returned by actor tasks from any plasma
+            # manager. Although the objects may still exist in that object
+            # store, this deletion makes them effectively unreachable by any
+            # local scheduler connected to a different store.
+            # TODO(swang): Actually remove the objects from the object store,
+            # so that the reconstructed actor can reuse the same object store.
+            if hex_to_binary(task["TaskSpec"]["ActorID"]) != NIL_ACTOR_ID:
+                dummy_object_id = task["TaskSpec"]["ReturnObjectIDs"][-1]
+                obj = self.state.object_table(dummy_object_id)
+                manager_ids = obj["ManagerIDs"]
+                if manager_ids is not None:
+                    # The dummy object should exist on at most one plasma
+                    # manager, the manager associated with the local scheduler
+                    # that died.
+                    assert(len(manager_ids) <= 1)
+                    for manager in manager_ids:
+                        # If the object was on a dead plasma manager, remove that
```

Comment:
Hmm.. the plasma manager will often not be dead (e.g., if the local scheduler dies it does not trigger the death of the manager). Wouldn't it make sense to go ahead and remove the object from the manager *that corresponds to the dead local scheduler*?

In our tests the manager will probably be dead because we kill the object store manually. But then the dummy objects are probably getting cleaned up by `cleanup_object_table` and not here.

So I have a feeling that this code isn't really necessary. Either the manager dies, in which case this happens anyway in `cleanup_object_table`, or the manager doesn't die, in which case this code won't do anything.

Does that sound right?

Reply:
This block is to handle the case that you wrote where the manager doesn't die. It removes the object from _any_ manager that had the object. For actor objects, the list should either have no managers (the object wasn't created and/or published yet), or the list should have one manager, in which case it will be the manager corresponding to the local scheduler. We have to remove the one manager so that reconstruction will go through.

This code does need to be replaced, though. Eventually, the plasma manager should be listening for a corresponding local scheduler's death, and remove the actor objects from the object table itself, rather than doing it in the monitor. Otherwise, there is a risk of a race here between the monitor removing the plasma manager and the plasma manager adding itself back (see the TODO at line 153).

Reply:
That makes sense. I find the comment confusing then. The comment says

```python
# If the object was on a dead plasma manager, remove that
# location entry.
```

Since it's possible that the plasma manager is still running, right? Is there a better way to phrase this comment?

Reply:
Oops, that was a comment left over from the copy pasta :) I'll update the comment.